File: examples/graduation_verdict/oracle.py
"""A second, independent, verdict-free implementation of the graduation decision.

This is deliberately the "naive way" from docs/architecture.md, generalized
to score *any* policy list rather than the fixed 7-subject curriculum — its
entire job is to be obviously correct by inspection, so it can serve as
ground truth for test_chaos.py's differential testing. See that file's own
docstring and docs/maintenance.md for the full reasoning: two independently
written implementations (this plain loop, and the verdict-based engine in
graduation_verdict.py) must agree on every input, or one of them is wrong.

Never import from `verdict` here — an oracle that shares a bug with the
system it's checking proves nothing.
"""

from __future__ import annotations

from graduation_verdict import SubjectPolicy


def _subject_passes(policy: SubjectPolicy, context: dict) -> bool:
    """Whether one subject's own scores clear its policy's bar — plain if/else, no Rule involved."""
    scores = context["scores"][policy.subject_id]

    if policy.subject_type == "vocational":
        return (
            scores["written_pct"] >= policy.written_min_pct
            and scores["practical_pct"] >= policy.practical_min_pct
        )

    if policy.subject_type == "language" and policy.exemption_allowed:
        return scores["written_pct"] >= policy.written_min_pct or scores.get("has_exemption", False)

    # academic, or a language subject with no exemption path
    return scores["written_pct"] >= policy.written_min_pct


def expected_graduates(policies: list[SubjectPolicy], context: dict, elective_minimum: int) -> bool:
    """Compute the graduation decision directly, with no `verdict` involved at all.

    Args:
        policies: Every subject's own policy (core and elective alike).
        context: One student's scores, cgpa, and attendance — the same
            shape `graduation_verdict.build_graduation_check`'s composite
            expects.
        elective_minimum: How many electives must pass.

    Returns:
        Whether this student graduates, computed independently of
        `rule_for_subject`/`build_graduation_check` — the two must always
        agree.
    """
    core_policies = [p for p in policies if not p.is_elective]
    elective_policies = [p for p in policies if p.is_elective]

    if not all(_subject_passes(p, context) for p in core_policies):
        return False

    passed_electives = sum(1 for p in elective_policies if _subject_passes(p, context))
    if passed_electives < elective_minimum:
        return False

    if context["cgpa"] < context["cgpa_floor"]:
        return False
    if context["attendance_pct"] < context["attendance_floor"]:
        return False

    return True

File: dart/examples/graduation_verdict/lib/src/oracle.dart
/// A second, independent, verdict_rules-free implementation of the graduation decision.
///
/// This is deliberately the "naive way" from
/// docs/samples/graduation-requirement-verdict/README.md, generalized to
/// score *any* policy list rather than the fixed 7-subject curriculum --
/// its entire job is to be obviously correct by inspection, so it can
/// serve as ground truth for the chaos suite's differential testing. See
/// that suite's own doc comment and docs/testing/README.md for the full
/// reasoning: two independently written implementations (this plain loop,
/// and the verdict_rules-based engine in graduation_check.dart) must agree
/// on every input, or one of them is wrong.
///
/// Never import package:verdict_rules here -- an oracle that shares a bug
/// with the system it's checking proves nothing.
library;

import 'subject_policy.dart';

/// Whether one subject's own scores clear its policy's bar -- plain
/// if/else, no rule involved.
bool _subjectPasses(SubjectPolicy policy, Map<String, Object?> context) {
  final scores = context['scores'] as Map<String, Object?>;
  final subject = scores[policy.subjectId] as Map<String, Object?>;
  final writtenPct = subject['written_pct'] as double;

  if (policy.subjectType == 'vocational') {
    final practicalPct = subject['practical_pct'] as double;
    return writtenPct >= policy.writtenMinPct &&
        practicalPct >= policy.practicalMinPct!;
  }

  if (policy.subjectType == 'language' && policy.exemptionAllowed) {
    final hasExemption = subject['has_exemption'] as bool? ?? false;
    return writtenPct >= policy.writtenMinPct || hasExemption;
  }

  // academic, or a language subject with no exemption path
  return writtenPct >= policy.writtenMinPct;
}

/// Compute the graduation decision directly, with no verdict_rules involved at all.
///
/// Returns whether this student graduates, computed independently of
/// `ruleForSubject`/`buildGraduationCheck` -- the two must always agree.
bool expectedGraduates(List<SubjectPolicy> policies,
    Map<String, Object?> context, int electiveMinimum) {
  final corePolicies = policies.where((p) => !p.isElective).toList();
  final electivePolicies = policies.where((p) => p.isElective).toList();

  if (!corePolicies.every((p) => _subjectPasses(p, context))) {
    return false;
  }

  final passedElectives =
      electivePolicies.where((p) => _subjectPasses(p, context)).length;
  if (passedElectives < electiveMinimum) {
    return false;
  }

  if ((context['cgpa'] as double) < (context['cgpa_floor'] as double))
    return false;
  if ((context['attendance_pct'] as double) <
      (context['attendance_floor'] as double)) return false;

  return true;
}
